rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,238 @@
+ff26cb9e-e034-4931-9c59-71bac578f7a0-0. During marker conflict detection, we do not need to list all the partitions
+under ".temp". Listing marker files under $BasePath/.hoodie/.temp/instantTime/2022/07/01 and checking if fileID
+ff26cb9e-e034-4931-9c59-71bac578f7a0-0 existed or not are sufficient.
+
+##### SimpleDirectMarkerBasedEarlyConflictDetectionStrategy
+
+Compared with `SimpleTransactionDirectMarkerBasedEarlyConflictDetectionStrategy`, this strategy drops the steps of
+locking including new transaction, begin transaction and end transaction. The advantages are that the checking speed is
+faster, and multiple writers do not affect each other. The downside is that the conflict detection may be delayed till
+pre-commit stage.
+
+For example, when two writers detect PartitionPath1 + fileID1 at the same time, both writers pass conflict detection and
+successfully create a marker, in which conflict detection is delayed. If so, the conflict can only be found in the
+pre-commit conflict detection and fail writing. This leads to waste of resources, but don't compromise the correctness
+of the data.
+
+As we can see, all these direct based early conflict detection strategy need extra FS calls.
+
+#### TimelineServerBasedWriteMarkers related strategy
+
+##### AsyncTimelineMarkerEarlyConflictDetectionStrategy
+
+This design expands the create marker API on timeline server.
+
+![](figure3.png)
+
+As shown in Figure 3, the client side calls create marker API, requesting to create markers.
+
+On the timeline server side, there is a MarkerCheckerRunnable thread, which checks the conflicts between the markers in
+memory created by the current writer and the markers persisted to the storage created by other active writers
+periodically. If any conflict is detected, this thread updates the value of hasConflict, which is used to respond to the
+marker creation requests.
+
+We did a performance test for async conflict checker using local FS and two concurrency writers based on 1500+ inflight
+markers located in MARKER0~MARKER19. It takes 179ms to finish checking conflict.
+
+Here we need to pay attention to the following details:
+
+Firstly, There may be a concurrency issue between reading and writing the same MARKER file causing EOFException. In
+order to simplify the model, we return empty result of markers when catching any exception during reading specific
+MARKER and let the subsequent checker to make up for it. For example, writer1 reads writer2's MARKER0 file, but at this
+moment writer2 is overwriting MARKER0 file. Finally, we get empty result after writer1 reading writer2's MARKER0. In
+extreme cases, the delay of early conflict detection will also occur. Fortunately, we can adjust the frequency of the
+schedule according to the actual situation.
+
+Secondly, the marker checker runnable only looks at the instants smaller than the current writer for conflict detection.
```

Comment:
It is not clear what this means to me. For every single write - Do we not have to check with all the existing markers by all the active writers to detect a conflict?. 

It makes sense to have some ordering to break ties between concurrent writes trying to create markers - but that is not what we are doing here though if I understand this right.

Reply:
The design here is to prevent two writers from detecting each other and all failing both.
For example writer1 start at 9:00 and writer2 start at 10:00 which conflict each other.

Without this limit, writer1 detect conflict with writer2 and failed. On the other hand writer2 also detect this conflict and failed either.

With this limit, only writer2 failed.

Reply:
Also we can design a priority strategy as follow up :)